db-engine/src/test/java/com/splicemachine/db/iapi/sql/dictionary/foreignkey/GraphTest.java

```diff
@@ -0,0 +1,239 @@
+            parameters.add(new Object[]{graph});
+            List<String> validGraphPermutations = permute(graph);
+            for (String permutedGraph : validGraphPermutations) {
+                parameters.add(new Object[]{permutedGraph});
+            }
+        }
+
+        @Parameterized.Parameters(name = "positive case --- {0}")
+        public static Collection<Object[]> data() throws StandardException {
+            addCase("digraph {\n" +
+                            "A -> B[label=\"Cascade\"];\n" +
+                            "A -> B[label=\"Cascade\"];\n" +
+                            "}");
+            addCase("digraph {\n" +
+                            "B -> A[label=\"Cascade\"];\n" +
+                            "C -> A[label=\"Cascade\"];\n" +
+                            "D -> B[label=\"Cascade\"];\n" +
+                            "D -> C[label=\"Cascade\"];\n" +
+                            "}");
+            addCase("digraph {\n" +
+                            "A -> B[label=\"SetNull\"];\n" +
+                            "C -> A[label=\"Cascade\"];\n" +
+                            "B -> C[label=\"NoAction\"];\n" +
+                            "}");
+            addCase("digraph {\n" +
+                            "D -> A[label=\"SetNull\"];\n" +
+                            "D -> C[label=\"Cascade\"];\n" +
+                            "C -> B[label=\"Restrict\"];\n" +
+                            "B -> A[label=\"Cascade\"];\n" +
+                            "}");
+            return parameters;
+        }
+
+        public PositiveTests(String graph) {
+            this.graph = graph;
+        }
+
+        String graph;
+
+        @Test
+        public void testcase() throws Exception {
+            Graph g = new GraphvizParser(graph).generateGraph("something");
+            GraphAnnotator annotator = new GraphAnnotator("something", g);
+            annotator.annotate();
+            annotator.analyzeAnnotations();
+        }
+    }
+
+    public static class NonParameterizedTests {
+        @Test
+        public void invalidGraphIsDetectedProperlyCase5() { // invalid cascade cycle
+            try {
+                new GraphvizParser
+                        ("digraph {\n" +
+                                 "A -> B[label=\"Cascade\"];\n" +
+                                 "B -> C[label=\"Cascade\"];\n" +
+                                 "C -> A[label=\"Cascade\"];\n" +
+                                 "}").generateGraph("something");
+                Assert.fail("expected graph creation to fail with error message: adding the constraint between A and C would cause the following " +
+                                    "illegal delete action cascade cycle A B C");
+            } catch (Exception e) {
+                Assert.assertTrue(e instanceof StandardException);
+                StandardException sqlException = (StandardException) e;
+                Assert.assertEquals("42915", sqlException.getSQLState());
```

Comment:
Maybe use LANG_DELETE_RULE_VIOLATION instead of 42915.

Reply:
Thanks! will keep that in mind when adding more fixes to FKs.